codegen/src/api/calls.rs

```diff
@@ -74,25 +76,31 @@ pub fn generate_calls(
                 };
 
             let pallet_name = &pallet.name;
-            let call_struct_name = &struct_def.name;
-            let function_name = struct_def.name.to_string().to_snake_case();
-            let fn_name = format_ident!("{}", function_name);
+            let call_name = variant_name;
+            let struct_name = &struct_def.name;
+            let call_hash = subxt_metadata::get_call_hash(metadata, pallet_name, call_name)
+                .unwrap_or_else(|_| abort_call_site!("Metadata information for the call {}_{} could not be found", pallet_name, call_name));
 
+            let fn_name = format_ident!("{}", variant_name.to_snake_case());
             let call_struct = quote! {
                 #struct_def
 
-                impl ::subxt::Call for #call_struct_name {
+                impl ::subxt::Call for #struct_name {
                     const PALLET: &'static str = #pallet_name;
-                    const FUNCTION: &'static str = #function_name;
+                    const FUNCTION: &'static str = #call_name;
                 }
             };
             let client_fn = quote! {
                 pub fn #fn_name(
                     &self,
                     #( #call_fn_args, )*
-                ) -> ::subxt::SubmittableExtrinsic<'a, T, X, #call_struct_name, DispatchError, root_mod::Event> {
-                    let call = #call_struct_name { #( #call_args, )* };
-                    ::subxt::SubmittableExtrinsic::new(self.client, call)
+                ) -> Result<::subxt::SubmittableExtrinsic<'a, T, X, #struct_name, DispatchError, root_mod::Event>, ::subxt::BasicError> {
+                    if self.client.metadata().call_hash::<#struct_name>()? == [#(#call_hash,)*] {
+                        let call = #struct_name { #( #call_args, )* };
+                        Ok(::subxt::SubmittableExtrinsic::new(self.client, call))
+                    } else {
+                        Err(::subxt::MetadataError::IncompatibleMetadata.into())
```

Comment:
I would like to see an integration test which tests this error condition, from what I can see all integration tests are just testing the success condition.

Reply:
Added in the latest revision